packages/@react-stately/utils/src/useControlledState.ts

```diff
@@ -13,11 +13,11 @@
 import {useCallback, useRef, useState} from 'react';
 
 export function useControlledState<T>(
-  value: T,
-  defaultValue: T,
-  onChange: (value: T, ...args: any[]) => void
-): [T, (value: T, ...args: any[]) => void]  {
-  let [stateValue, setStateValue] = useState(value || defaultValue);
+  value?: T | null,
+  defaultValue?: T | null,
+  onChange?: (value: T | undefined | null, ...args: any[]) => void
```

Comment:
It may depend on the component whether the `onChange` value can be null/undefined - I think in many cases it cannot? And what about the value/defaultValue - are those optional in all cases? Should we instead change the uses of `useControlledState` so that `T` contains `undefined | null` only if needed? Maybe a second generic parameter (with a default) if we need a different type for onChange vs value?

Reply:
Ah, I see about null, users can provide that `useControlledState<string | null>(null, undefined, console.log)` since it can't occur unless they specify it through the setState, value, or defaultValue

so unfortunately the hook can be used both of these ways
```
useControlledState(undefined, 'default', console.log)
useControlledState('value', undefined, console.log)
```
So either I create two override interfaces where value or defaultValue is required, or they have to both be optional from what I understand.

I don't quite follow what the second generic would be for?

Reply:
I guess writing out `| undefined` explicitly isn't necessary since you already made the parameters optional. But the value passed to onChange shouldn't allow undefined/null unless `T` provided by the caller does.

https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABCAzgUwMIKgJzgG3zQBMBlKAQyjQB4AVAGkQ0QF5E6A+ACgDcL8INAH4AXBybE0wCiHxQAagKFiJiBBgAWFMAHMR4vuIwBKNp0S84MYmYDeAXwBQTgPSvmiGCkRg0vNBxkMClgGD9iADonVExsPEIScio0bgByYDg4NKZwUPCSJj4zVgtHEwBuGPQsMFwCIjJKam486QLiJjSAIwocHMRi80RyqrcPAFEADwAHfBgIGCh8AE9EAXw4AHdfOXxEKDhEGbwZpm6QKEQAWwoAazR1MC0dfXX8LYoVn27HlFxwrporFavVEk0UjR-jhAYgAD67QhMaGAngZLIDNphCJFXglMoOSrVOJ1BKNZLUKEAvTwxH4ZHU3Q8LEdLq9fq4-EjQljIA

Reply:
Yeah, I went through this at one point too. However, I think there is a problem with it. Let me know if I'm understanding this correctly.

The issue arises though with the returned value of the useControlledState
see https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwK4GcCmAlNAzANHVNAZRgEMY0BfObKCEOAcijVIGMZGBuAKB+yQA7DsAiCC6AMJiYdADZy0AExLk0AHgAq+SXAC8cTQD4AFADdScpGgD8ALkP4lOUkjkwAapev3HcMZIAFqSCAOa2DuYOkgCU+kZwZhDASjEOANracObeaA5ZAHRFpFChKA4hAJ7pALpxeglJKTWIPHDtcIrw6ShkFF5WaPjoMKr9uS0GhGNoOYNwAD4LcM7Yru4D1jFtHV1wvWpY2PoSmDgmB+ODMbwdnWjwwCjSgrIQCsonFvMAhHpTglWwEEyluHWAxxMTxebw+qVadzulzO2AKbCQUBYry+uTB7WoaDk6ARiPa32sJ2RmzQeLglB27RYMAx4nS5KG+weM2pNV49J4AHoBXBdE84CCzGgoARAThgcoCjxCDD5IoVH1ZoxsBAIIx8EIgSClPhzPUEghKDclVIZKrlDMTAa5Ub8IwAEYlPXZMxmxCW3iC4UAUQAHmA5MA2MAYHJKnBLHIIAB3cVuORwGAQOBgOhgfBupDwECkADWaH8giCIXC8bkSdIlRQcDd5d6UGBoUVytt7zVM3UbY7i1TCmGsg7pi1Oq9Tuw8uN3t9Fqt3dedvVagH47Cw8EabH7bCpln89dHqgXtN8T9Vp4QA

Technically, the user may set something that always returns a non-undefined value. However, because both the input value and defaultValue are optional, the type for the returned value must always include undefined, even if it's impossible.

Reply:
You could use overloads for that:

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwK4GcCmAlNAzANHVNAZRgEMY0BfObKCEOAcijVIGMZGBuAKB+yQA7DsAiCC6AMJiYdADZy0AExLk0AHgAq+SXAC8cTQD4AFADdScpGgBch-EpykkcmADVL1u5rgAfAoKO2MCCyvhikgAWpIIA5mgA-HbmdpIAlPpGcGYQwEppdgDa2nDmnrb2cAB0NaRQsSh2MQCehQC6GXpZOXltvALCMKLihNKCshAKyqoUWjr6hqYWVhU+-kJBIWFwQc6uHive4YJRMfFJpWapnd25+UUlZYeVNVV1DU2CrR2Z2Xd9-CEIjEEjQYwmUxUZFmJV0BmMTy8lV2Lnc5SOcAi0TiiWSVzg6V+PXucGK+ERq3wr3ejTgLXaNz+vUQPDgbLginghRQ0LQB2s+HQMBmfPKbQWhBFFL8-hR+3KaVZ7M5cB5aiw2Al6A1JjVFH5aDSvHZHLQ8GAKHB8kUSgWy2scAAhHoDBscFslMb2cBNSYLVbJjaMgglSa2XrMDgqmwkFAWOM7eUvWzqGg5OgWWH2fa0AsIwbk3BKKG4CwYLHxIUc4KzSKDQDizwAPRNglwC1wUJmNBQAKbUJKKo8UYya3TXkmRjYCAQRj4N3BAfksyMhCUI3DqSjwPjtQmBce-CMABGdTnl1X694zdbAFEAB5gOTANjAGByZp0hQQADunZcchwDAEBwGAdBgPgx5IPAICkAA1rmWJnLmlg-qQzQoHAx65jyUAhLEQ4juMY5Qmo6i4fhfj-gogqyPhphTjO54HkuF6-GuG5ERCNoiuRdFxFRggAbReFxKYLHbCeZ7LpeG43nAABCEAwJEcBsDEgjKVhuYSYOm5gtukK8RRAn+EJNGqvxsTiYE7qsbpMnsVePBAA

This way only one or the other can be undefined, not both.